File: app/sql/sqlwrite.py
from sql.demo_patterns import SAMPLE_PATTERNS


SQL_BUILD_INSERTION = """
INSERT INTO builds(build_num, vcs_revision, queued_at, job_name)
VALUES(%s, %s, %s, %s)
ON CONFLICT (build_num) DO NOTHING;
"""


# XXX Keep this list up to date so that "scrubbing" the database
# catches everything.
TABLE_NAMES = [
    "scanned_patterns",
    "scans",
    "matches",
    "build_steps",
    "builds",
    "pattern_tags",
    "patterns",
]


def insert_builds(conn, values_to_insert):

    with conn.cursor() as cur:
        for vals in values_to_insert:
            cur.execute(SQL_BUILD_INSERTION, vals)

        conn.commit()


match_insertion_sql = """
INSERT INTO matches(build_step, pattern, line_number, line_text, span_start, span_end)
VALUES(%s, %s, %s, %s, %s, %s);
"""


build_scan_record_insertion_sql = """
INSERT INTO scanned_patterns(build, scan, pattern)
VALUES(%s, %s, %s);
"""


def insert_matches(engine, results):

    results_list = list(results)
    with engine.conn.cursor() as cur:

        # There can be en entry for a build_num even if it had no matches.
        for i, (fetch_success, build_num, searched_pattern_ids, scan_id, build_step_data_tuples) in enumerate(results_list):

            engine.logger.log("On %d/%d results..." % (i, len(results_list)))

            if not fetch_success:
                engine.logger.log("WARNING: Skipping unsuccessful log request for build %d..." % build_num)
                continue

            # NOTE: There will be at most one iteration of this loop, because it is always a singleton list upon success
            for j, (step_name, is_timeout, (line_count, match_list)) in enumerate(build_step_data_tuples):

                engine.logger.log("\tOn %d/%d build_step_data_tuples..." % (j, len(build_step_data_tuples)))

                step_id = insert_build_step(engine.conn, build_num, step_name, is_timeout)

                for k, (pattern_id, line_number, line_text, match_span) in enumerate(match_list):

                    engine.logger.log("\t\tOn %d/%d match_list..." % (k, len(match_list)))

                    match_start, match_end = match_span
                    vals = (step_id, pattern_id, line_number, line_text, match_start, match_end)
                    cur.execute(match_insertion_sql, vals)


            # FIXME represent this differently
            # for j, patt_id in enumerate(searched_pattern_ids):
            #
            #     engine.logger.log("\tXXXX On %d/%d searched_pattern_ids..." % (j, len(searched_pattern_ids)))
            #
            #     build_scan_vals = (build_num, scan_id, patt_id)
            #     cur.execute(build_scan_record_insertion_sql, build_scan_vals)

        engine.conn.commit()


def scrub_tables(conn):
    with conn.cursor() as cur:
        for table in TABLE_NAMES:
            cur.execute("TRUNCATE %s CASCADE;" % table)


def insert_for_id(conn, sql, args):
    with conn.cursor() as cur:
        cur.execute(sql, args)
        conn.commit()
        return cur.fetchone()[0]


def insert_scan_row(conn):
    insertion_sql = """
    INSERT INTO scans(timestamp)
    VALUES(now()) RETURNING id;
    """
    return insert_for_id(conn, insertion_sql, ())


def insert_build_step(conn, build_id, name, is_timeout):
    insertion_sql = """
    INSERT INTO build_steps(build, name, is_timeout)
    VALUES(%s, %s, %s) RETURNING id;
    """
    return insert_for_id(conn, insertion_sql, (build_id, name, is_timeout))


def populate_patterns(conn):

    pattern_insertion_sql = """
    INSERT INTO patterns(regex, expression, description, is_infra) VALUES(%s, %s, %s, %s) RETURNING id;
    """

    tag_insertion_sql = """
    INSERT INTO pattern_tags(tag, pattern) VALUES(%s, %s);
    """

    applicable_step_insertion_sql = """
    INSERT INTO pattern_step_applicability(step_name, pattern) VALUES(%s, %s);
    """

    with conn.cursor() as cur:

        for is_regex, sample_pattern, description, is_infra, tags, applicable_steps in SAMPLE_PATTERNS:
            cur.execute(pattern_insertion_sql, (is_regex, sample_pattern, description, is_infra))
            pattern_id = cur.fetchone()[0]

            for tag in tags:
                cur.execute(tag_insertion_sql, (tag, pattern_id))

            for applicable_step in applicable_steps:
                cur.execute(applicable_step_insertion_sql, (applicable_step, pattern_id))

        conn.commit()

File: haskell/fetcher/src/SqlWrite.hs
{-# LANGUAGE OverloadedStrings #-}

module SqlWrite where

import           Builds
import           Data.Foldable              (for_)
import qualified Data.Maybe                 as Maybe
import           Data.Text                  (Text)
import qualified Data.Text                  as T
import qualified Data.Text.IO               as TIO
import           Data.Time.Format           (defaultTimeLocale, formatTime,
                                             rfc822DateFormat)
import           Database.PostgreSQL.Simple
import           GHC.Int                    (Int64)

import qualified DbHelpers
import qualified ScanPatterns
import qualified ScanRecords
import qualified ScanUtils


-- | We do not wipe the "builds" or "build_steps" tables
-- because visiting each build is expensive.
allTableTruncations :: [Query]
allTableTruncations = [
    "TRUNCATE scanned_patterns CASCADE;"
  , "TRUNCATE scans CASCADE;"
  , "TRUNCATE matches CASCADE;"
  , "TRUNCATE pattern_step_applicability CASCADE;"
  , "TRUNCATE pattern_tags CASCADE;"
  , "TRUNCATE patterns CASCADE;"
  , "TRUNCATE log_metadata CASCADE;"
  , "TRUNCATE build_steps CASCADE;"
  , "TRUNCATE builds CASCADE;"
  ]


prepare_database :: IO Connection
prepare_database = do

  conn <- DbHelpers.get_connection
--  SqlWrite.scrub_tables conn
  SqlWrite.populate_patterns conn ScanPatterns.pattern_list
  return conn


scrub_tables :: Connection -> IO ()
scrub_tables conn = do

  for_ table_truncation_commands $ \table_truncation_command -> do
    execute_ conn table_truncation_command

  where
    table_truncation_commands = allTableTruncations

    -- TODO optionally exclude the last table, which is the "builds" table
    -- table_truncation_commands = init allTableTruncations


build_to_tuple :: Build -> (Int64, Text, Text, Text, Text)
build_to_tuple (NewBuild (NewBuildNumber build_num) vcs_rev queuedat jobname branch) = (build_num, vcs_rev, queued_at_string, jobname, branch)
  where
    queued_at_string = T.pack $ formatTime defaultTimeLocale rfc822DateFormat queuedat


store_builds_list :: Connection -> [Build] -> IO Int64
store_builds_list conn builds_list =

  executeMany conn "INSERT INTO builds(build_num, vcs_revision, queued_at, job_name, branch) VALUES(?,?,?,?,?) ON CONFLICT (build_num) DO NOTHING" $
    map build_to_tuple builds_list


store_matches :: ScanRecords.ScanCatchupResources -> BuildStepId -> BuildNumber -> [ScanPatterns.ScanMatch] -> IO Int64
store_matches scan_resources (NewBuildStepId build_step_id) _build_num scoped_matches =
  executeMany conn insertion_sql $ map to_tuple scoped_matches

  where
    conn = ScanRecords.db_conn $ ScanRecords.fetching scan_resources
    scan_id = ScanRecords.scan_id scan_resources

    to_tuple match = (
        scan_id
      , build_step_id
      , DbHelpers.db_id $ ScanPatterns.scanned_pattern match
      , ScanPatterns.line_number match_deets
      , ScanPatterns.line_text match_deets
      , ScanPatterns.start $ ScanPatterns.span match_deets
      , ScanPatterns.end $ ScanPatterns.span match_deets
      )
      where
        match_deets = ScanPatterns.match_details match

    insertion_sql = "INSERT INTO matches(scan_id, build_step, pattern, line_number, line_text, span_start, span_end) VALUES(?,?,?,?,?,?,?);"


insert_single_pattern :: Connection -> ScanPatterns.Pattern -> IO Int64
insert_single_pattern conn (ScanPatterns.NewPattern expression_obj description tags applicable_steps) = do

  [Only pattern_id] <- query conn pattern_insertion_sql (ScanPatterns.is_regex expression_obj, ScanPatterns.pattern_text expression_obj, description, False, False)

  for_ tags $ \tag -> do
    execute conn tag_insertion_sql (tag, pattern_id)

  for_ applicable_steps $ \applicable_step ->
    execute conn applicable_step_insertion_sql (applicable_step, pattern_id)

  return pattern_id

  where
    pattern_insertion_sql = "INSERT INTO patterns(regex, expression, description, is_infra, has_nondeterministic_values) VALUES(?,?,?,?,?) RETURNING id;"
    tag_insertion_sql = "INSERT INTO pattern_tags(tag, pattern) VALUES(?,?);"
    applicable_step_insertion_sql = "INSERT INTO pattern_step_applicability(step_name, pattern) VALUES(?,?);"


populate_patterns :: Connection -> [ScanPatterns.Pattern] -> IO ()
populate_patterns conn pattern_list = do
  for_ pattern_list $ insert_single_pattern conn


step_failure_to_tuple :: (BuildNumber, Either BuildStepFailure ScanRecords.UnidentifiedBuildFailure) -> (Int64, Maybe Text, Bool)
step_failure_to_tuple (NewBuildNumber buildnum, visitation_result) = case visitation_result of
  Right _ -> (buildnum, Nothing, False)
  Left (NewBuildStepFailure stepname mode) -> let
    is_timeout = case mode of
      BuildTimeoutFailure              -> True
      ScannableFailure _failure_output -> False
    in (buildnum, Just stepname, is_timeout)


store_log_info :: ScanRecords.ScanCatchupResources -> BuildStepId -> ScanRecords.LogInfo -> IO Int64
store_log_info scan_resources (NewBuildStepId step_id) (ScanRecords.LogInfo byte_count line_count) = do

  execute conn "INSERT INTO log_metadata(step, line_count, byte_count) VALUES(?,?,?);" (step_id, line_count, byte_count)

  where
    conn = ScanRecords.db_conn $ ScanRecords.fetching scan_resources


insert_build_visitation :: ScanRecords.ScanCatchupResources -> (BuildNumber, Either BuildStepFailure ScanRecords.UnidentifiedBuildFailure) -> IO BuildStepId
insert_build_visitation scan_resources visitation = do

  [Only step_id] <- query conn "INSERT INTO build_steps(build, name, is_timeout) VALUES(?,?,?) RETURNING id;" $ step_failure_to_tuple visitation
  return $ NewBuildStepId step_id

  where
    conn = ScanRecords.db_conn $ ScanRecords.fetching scan_resources


insert_scan_id :: Connection -> ScanRecords.PatternId -> IO Int64
insert_scan_id conn (ScanRecords.NewPatternId pattern_id)  = do
  [Only pattern_id] <- query conn "INSERT INTO scans(latest_pattern_id) VALUES(?) RETURNING id;" (Only pattern_id)
  return pattern_id


-- TODO finish this
api_new_pattern_test :: Builds.BuildNumber -> ScanPatterns.Pattern -> IO [ScanPatterns.ScanMatch]
api_new_pattern_test build_number new_pattern = do

  full_filepath <- ScanUtils.gen_log_path build_number
  putStrLn $ "Scanning log: " ++ full_filepath

  -- TODO consolidate with Scanning.scan_log
  console_log <- TIO.readFile full_filepath
  let lines_list = T.lines console_log
      result = Maybe.mapMaybe apply_pattern $ zip [0::Int ..] $ map T.stripEnd lines_list

  putStrLn $ "Results: " ++ show (length result)

  return result

  where

    apply_pattern :: (Int, Text) -> Maybe ScanPatterns.ScanMatch
    apply_pattern line_tuple = ScanUtils.apply_single_pattern line_tuple $ DbHelpers.WithId 0 new_pattern



-- TODO not used yet
api_new_pattern :: ScanPatterns.Pattern -> IO Int64
api_new_pattern new_pattern = do
  conn <- DbHelpers.get_connection
  insert_single_pattern conn new_pattern
